### ISSUE TITLE
JavaScript Widgets and Accessibility: which browser/screenreader combinations should be ok?

### ISSUE THREAD

**Issue by jmuheim** [twbs/bootstrap#18004]
I'm working for an accessibility consultancy in Switzerland, [Access for all](http://www.access-for-all.ch). We are very often asked which JavaScript widgets we would advise to use for implementing accessible websites.

I've done quite some research about accessible tab panels, and the conclusion so far is: none of the tested implementations are fully satisfying in the newest versions of NVDA/JAWS, running on Firefox/IE. So this is quite a pity. (I will publish my results some time soon and link it here.)

Personally, I love Bootstrap a lot, and while its JavaScript widgets are pretty well done and working in NVDA, they make problems in JAWS: for example, the [accordion](http://getbootstrap.com/javascript/#collapse-example-accordion) works nicely in IE, but it absolutely doesn't work in FF.

I love how simplistic the ARIA implementations of Bootstrap are. In fact: while ARIA is a good invention, browser support is horrible across operating systems on both desktop and mobile, so I recommend my clients to avoid it in most respects and stick to traditional HTML constructs instead, enhanced with some simple JavaScript. But with the simplistic implementations of Bootstrap widgets, I guess it should be possible to make them work in both NVDA and JAWS in the newest versions of both IE and Firefox.

**So I wonder: what browser/screenreader combinations is Bootstrap committed to support in respect of accessibility?**

It would be wonderful if I could recommend the Bootstrap JavaScript widgets without restriction to all my clients.

PS: I asked a similar question about the carousel on StackOverflow: [Widgets working nicely on NVDA, but JAWS? Accordion doesn't work](http://stackoverflow.com/questions/31029502/widgets-working-nicely-on-nvda-but-jaws-accordion-doesnt-work).

**Comment by patrickhlauke** [comment]
There are many known issues with some of our more complex widgets. Ideally, we should be supporting most common combinations: Safari/VoiceOver on iOS and OS X, Chrome/TalkBack (within the limits of what's possible, as that combination exposes very little in terms of semantics and ARIA to the user) and Firefox/TalkBack on Android; all combinations of IE/Firefox/Chrome and NVDA/JAWS on Windows. Unfortunately, there's a lot of work to do on some of the widgets to get it up to this ideal. I was actually unaware of the issue with Firefox/JAWS, and to my knowledge nobody had reported any issues to that effect here. So, as a starting point, I'd be delighted to see specific issues (and, even better, PRs) being filed here for these sorts of things. Ideally, we'd want to iron out most glaring problems with widgets by the time v4 is released.

**Comment by kkirsche** [comment]
Hey @patrickhlauke do you have any guides on how to set these up locally to test? I don't mind helping out but don't know where these features are to test.

**Comment by patrickhlauke** [comment]
One of the problems here is that to test with screen readers and get some solid test results, you do need to be fairly well versed in screen reader usage, and know what the expected output/behavior should be. That's part of the reason why it's not often that we get solid bug reports.
In terms of setting up to test, you'll need the individual browsers and assistive technologies. Some are free (like NVDA for Win, VoiceOver being already included in OS X and iOS, ditto for TalkBack on Android), while others (JAWS) are expensive pieces of software...

**Comment by jmuheim** [comment]
Thanks for your reply, Patrick.

I'm very aware of the problems of screenreader testing on both desktop and mobile, as this is our daily business (we're counselling web development agencies each and every day in making their products more accessible).

I'd be very interested in offering you our help (I'm an experienced web developer in both frontend and backend, and I'm working with both handicapped and non-handicapped collegues), so maybe we could join forces?

**Comment by patrickhlauke** [comment]
@jmuheim my comment about difficulty of SR testing was directed at @kkirsche ;)

by all means any help would be appreciated.

**Comment by jmuheim** [comment]
> by all means any help would be appreciated.

I'm only brainstorming right now, but how could such a collaboration look like? Are you working with other people specifically on the topic of accessibility? Are there guidelines? How do you check back whether new results (e.g. in form of pull requests) meet the accessibility requirements?

**Comment by patrickhlauke** [comment]
at the moment what's needed is a gap analysis of all current widgets, documenting where they fall short of ideal.

if any of the changes entail different ARIA roles/attributes to be used, we then should ideally also get unit tests set up for those. these will ensure that future changes don't break what's there.

beyond that, I (and many others, like @cvrebert) keep an eye out for any possible accessibility issues that come in with new proposed PRs

**Comment by heikojansen** [comment]
For Bootstrap 3 there's https://github.com/paypal/bootstrap-accessibility-plugin but it hasn't seen many updates lately and as far as I can tell at least a few things dynamically provided by it are now regular parts of Bootstrap (v4).
Wouldn't it make sense to analyze it for further enhancements of v4? Or has that plugin been completely superseeded by v4 development?

**Comment by patrickhlauke** [comment]
a number of improvements have already been made in core bootstrap that make certain aspects of the paypal plugin redundant. there are also certain aspects of the plugin which we don't quite agree with (such as the keyboard handling for widgets like menus, which the plugin bases solely on the ARIA spec while we'd like to be a bit more holistic, allowing for TAB as well as cursor keys). but yes, generally it's good to crib from the paypal plugin to get inspiration of what could be modded straight into core.

**Comment by mdo** [comment]
Closing since I think the discussion was had and answers provided. Holler if we need to reopen.
